### ISSUE TITLE
TCP interface specification/selection

### ISSUE THREAD

**Issue by rhc54** [open-mpi/ompi#9682]
This is a general question arising from work on auto-translating OMPI RTE-related params to their PMIx/PRRTE versions. It particularly impacts people using containers, but it is something we have seen arise from the OMPI user community repeatedly over the years.

If the user specifies TCP networks to use - e.g., with `btl_tcp_if_include` - then should the RTE (including PMIx in that for now) also use those adaptors? Or does the user have to _know_ that OMPI has multiple layers using TCP and therefore has to set multiple MCA params to make the package work?

I know there are times we want the separate out the MPI from the RTE traffic, and times when you have no choice but to commingle the messaging due to connectivity limitations. This goes a bit to "work out of the box" versus "customize for performance" - not sure of the right compromise.

I assigned this question to @abouteiller last week - but #9681 reminded me that others (e.g., @jjhursey) are similarly impacted.

Anyone have input here?

**Comment by jjhursey** [comment]
> user specifies TCP networks to use - e.g., with btl_tcp_if_include - then should the RTE also use those adaptors?

No. Here they are directing a specific component to use these interfaces (`btl_tcp`). So they are being specific in their intentions. Having a carried side effect of changing other parameters would be unexpected.

Often we set the `btl_tcp_if_include` to the high-speed ethernet interfaces, and the `oob_tcp_if_include` to a single management interface so isolate traffic.

If you wanted to have a general `-mca if_include` option then I could see that setting the default for the `btl_tcp_if_include` / `oob_tcp_if_include` (and other places that want this information). Since it is no longer tied to a specific component (and is a new option) then this would probably fit better with the user's expectations.

**Comment by rhc54** [comment]
Okay, then it sounds like we should only translate _literally_ when dealing with the OMPI MCA parameters. I'm okay with doing that, but wanted to check the pulse of folks over here.

It means the users have to confront the PRRTE vs PMIx vs OMPI project splits - but that's only one step worse than what they had to do before. So in this case, if they needed to stipulate interfaces to enable connectivity, they'll have to figure out to set btl/tcp, oob/tcp, and ptl/tcp parameters.

Perhaps someone should document that in the FAQ? Otherwise, I'm not sure how a user would discover this on their own without considerable pain.

**Comment by jsquyres** [comment]
I agree with @jjhursey -- let's keep them separate for now.

Similar to what @jjhursey said, we might want to introduce a new OMPI MCA var that is just for IP interfaces.  I'd suggest something like `ip_if_include`, since we're talking about IP Linux networking interfaces.  Although `tcp_if_include` might be a bit more recognizable / palatable to end users (we could certainly make `tcp_if_include` be an alias of `ip_if_include`)

The idea would be similar to what has already been described: the common case is where a user only wants to have to set one MCA param for all OMPI/etc. IP traffic.  The user therefore doesn't have to know/care what subsystems / frameworks / components are impacted.  The other / legacy MCA IP interface selection MCA params are still there if finer-grained control is necessary.

**Comment by rhc54** [comment]
Sounds good. Only one corner-case I can think of: if someone specifies an oob/tcp param, I would think we'd want to set the corresponding ptl/tcp param - yes? Both relate to RTE functions, so it feels like the right think to do.

**Comment by jsquyres** [comment]
Maybe we should have a map of (old) OMPI MCA params to (new) PRTE MCA params?

**Comment by rhc54** [comment]
In a perfect world...yeah, someone could create a table in the FAQ area. Almost all of them directly come across. There are a few, though, that are a little tricky as there isn't a direct correlation.

I took a first pass in the translator code in PRRTE that tries to handle some of the exceptions, and created synonyms to help with others so this can be as transparent as possible. Not 100% though. I'm sure someone could improve it.

**Comment by abouteiller** [comment]
The argument that btl_tcp_xyz is very specific in its intention to address the BTL framework (and only it) is very potent. 

I would still strongly recommend that we do have an all-emcompassing option, as described above by Jeff with ip_if_include/tcp_if_include that can set a general default for all derived parameters (btl_tcp_, oob_tcp_, others...) and can still be overridden.
